### ISSUE TITLE
NPE from Netty4Transport#stopInternal if eventLoopGroup == null

### ISSUE THREAD

**Issue by trejkaz** [elastic/elasticsearch#56068]
**Elasticsearch version** (`bin/elasticsearch --version`): 6.8.0

**Plugins installed**: [parent-join-client, analysis-icu-client, lang-mustache-client, transport-netty4-client, percolator-client, reindex-client, x-pack-transport]

**JVM version** (`java -version`): 1.8.0_191 (Oracle)

**OS version** (`uname -a` if on a Unix-like system): Windows Server 2016

**Description of the problem including expected versus actual behavior**:

An end user reported getting failures to connect to Elasticsearch. I see no errors coming from the code which should be opening the transport, but I do see many errors about closing it:

```lang-none
java.lang.NullPointerException: null
	at org.elasticsearch.transport.netty4.Netty4Transport.lambda$stopInternal$2(Netty4Transport.java:255) ~[transport-netty4-client-6.8.0.jar:6.8.0]
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:103) ~[elasticsearch-core-6.8.0.jar:6.8.0]
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:85) ~[elasticsearch-core-6.8.0.jar:6.8.0]
	at org.elasticsearch.common.lease.Releasables.close(Releasables.java:36) ~[elasticsearch-6.8.0.jar:6.8.0]
	at org.elasticsearch.common.lease.Releasables.close(Releasables.java:46) ~[elasticsearch-6.8.0.jar:6.8.0]
	at org.elasticsearch.common.lease.Releasables.close(Releasables.java:51) ~[elasticsearch-6.8.0.jar:6.8.0]
	at org.elasticsearch.transport.netty4.Netty4Transport.stopInternal(Netty4Transport.java:254) ~[transport-netty4-client-6.8.0.jar:6.8.0]
	at org.elasticsearch.transport.TcpTransport.lambda$doStop$11(TcpTransport.java:578) ~[elasticsearch-6.8.0.jar:6.8.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681) ~[elasticsearch-6.8.0.jar:6.8.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_172]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_172]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_172]
```

On inspecting code in the area, I see:

```java
    @Override
    @SuppressForbidden(reason = "debug")
    protected void stopInternal() {
        Releasables.close(() -> {
            Future<?> shutdownFuture = eventLoopGroup.shutdownGracefully(0, 5, TimeUnit.SECONDS);
            // ... omitting rest of method ...
        });
    }
```

So `eventLoopGroup` must be `null` for some reason, but isn't being checked here.

I assume that under _normal_ conditions, it must _never_ be null during this method, but that something failed during startup in this case. However, there is nothing in the logs about an error starting the transport.

On further investigation, this turns out to be because doStop() is called from a finally block, so the exception from doStop() is guaranteed to clobber any exception which it's cleaning up after:

```java
    @Override
    protected void doStart() {
        boolean success = false;
        try {
            👇Exception is either thrown from here
            ThreadFactory threadFactory = daemonThreadFactory(settings, TRANSPORT_WORKER_THREAD_NAME_PREFIX);

            👇Or from here
            eventLoopGroup = new NioEventLoopGroup(workerCount, threadFactory);

            // ... omitting further startup code ...
            success = true;
        } finally {
            if (success == false) {
                doStop(); 👈but exception from here clobbers any stack traces we wanted
            }
        }
    }
```

**Steps to reproduce**:

The best idea I have at the moment is that I see us creating the client 258 times in the logs. Could there possibly be a hard limit of 256 clients in the same JVM, or on the same system? It seems crazy though, I'm used to macOS limiting things like that, but Windows?

**Provide logs (if relevant)**:

**Comment by original-brownbear** [comment]
@trejkaz I think the most likely cause of the problem that too many transport clients and hence transport services will create too many threads. Ideally you should simply not be creating that many transport clients and simply share them across your application to begin with.

That said, even though this is a bit of a fringe issue, we should probably still clean up the potential NPE to avoid swallowing any exceptions here like we do now so I marked this as a ">bug" for the time being.

**Comment by original-brownbear** [comment]
Fixed in #56080
